### PR TITLE
feat!: Add Discover, Metadata, User Info activities

### PR DESCRIPTION
**Pull request by rcooney**

### CODE REVIEW

File: src/activities/CartegraphSignIn.ts
Comment (cdoak-gh) [inline src/activities/CartegraphSignIn.ts:1]: Password is required but the password parameter is optional.

Comment (cdoak-gh) [inline src/activities/CartegraphSignIn.ts:1]: Password is required but is an optional parameter

File: src/activities/DeleteCartegraphRecord.ts [hunk 1]
```diff
@@ -62,7 +63,7 @@ export default class DeleteCartegraphRecord implements IActivityHandler {
         const url = new URL(
             `${service.url}/api/v1/classes/${encodeURIComponent(className)}`,
         );
-        if (id) {
+        if (typeof id === "number") {
             url.pathname += `/${id}`;
         }
         filter && url.searchParams.append("filter", filter);
```
Comment (cdoak-gh) [inline src/activities/DeleteCartegraphRecord.ts:69]: Do we want to urlEncode parameters or does searchParams take care of that for us?

File: src/activities/SendCartegraphRequest.ts
Comment (cdoak-gh) [inline src/activities/SendCartegraphRequest.ts:1]: const url = new URL(`${service.url}/${path}`);  
Should path be urlEncoded?

File: src/activities/CartegraphSignIn.ts
Comment (rcooney) [reply src/activities/CartegraphSignIn.ts:1]: Interesting. This is an area that we don't handle very consistently in our activity packs.

I think the best approach is to mark all activity inputs as optional in TypeScript regardless of the `@required` decoration they have. There is nothing to stop a workflow author from assigning an value that will be `undefined` at runtime to a `@required` input, so we should type our activity inputs to account for that.

I've just pushed an update to this PR that marks all inputs as optional.

File: src/activities/SendCartegraphRequest.ts
Comment (rcooney) [reply src/activities/SendCartegraphRequest.ts:1]: For an individual path segment I would normally say yes. However, this is a full path with forward slashes. We don't want those slashes encoded. And I don't think we want to try to pick apart their path.

**Review by cdoak-gh — APPROVED** [review]

**Pull request merged**